src/lib/operators/projection.cpp

```diff
@@ -226,7 +240,7 @@ std::shared_ptr<const Table> Projection::_on_execute() {
     const auto entire_chunk_pos_list = std::make_shared<EntireChunkPosList>(chunk_id, input_chunk->size());
     for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
       // Turn newly generated ValueSegments into ReferenceSegments, if needed
-      if (expressions[column_id]->type != ExpressionType::PQPColumn && output_table_type == TableType::References) {
+      if ((expressions[column_id]->type != ExpressionType::PQPColumn || pqp_columns_to_evaluate[column_id]) && output_table_type == TableType::References) {
```

Comment:
Personally, I feel like returning an `unordered_set<AbstractExpression> forwarded_columns`  and checking `forwarded_columns.contains(...)` would be more idiomatic.

Reply:
I don't know what it was exactly, but I had some problems with abstract expressions and hashing at some point. Do you think it's safe to do?

Could thinks happen like "two different PQPColumn Expressions that link the same actual column"?

Reply:
> Could thinks happen like "two different PQPColumn Expressions that link the same actual column"?

Two different PQPColumnExpression objects? Yes. But they are equi-comparable.

Use `ExpressionUnorderedSet`, not `unordered_set<shared_ptr<AbstractExpression>>`.

Reply:
Soooo  much better with `ExpressionUnorderedSet`!

Reply:
Probably should add a test (if you already haven't) that works with ExpressionUnorderedSet and not with the default unordered_set

Reply:
I am not sure what you mean by that (not using an `unordered_set` anywhere), but your  mention of `ExpressionUnorderedSet` made me add a new test `ExpressionUnorderedSetCheck`, which checks if the assumption that different PQPColumnExpressions referencing the same column are actually equal within the set.
They are.

Is that what you meant?

Reply:
I was thinking about a test in the list of projection tests that checks that with one forwarded expression that references one column and another nested expression that causes that column to be materialized, the correct thing is done, even if the expression *objects* are not identical. Maybe the tests already do that though.

Reply:
That should be the `ExpressionUnorderedSetCheck` test.

Reply:
Ah, sorry. I understood your last comment here as if the `ExpressionUnorderedSetCheck` was checking the `ExpressionUnorderedSet` itself, not its use in here.